1.21.1 RS Bridge to RS2 port

SirEndii opened the pull request:
Still needs a description - currently trying to use the new rs2 api to make the RS bridge
Low prio because RS2 isn't released yet

SirEndii committed:
RS2 RS Bridge without bridge

SirEndii committed:
RS Bridge can now connect to a RS System

SirEndii committed:
Use for the RS functions our `IStorageSystemPeripheral` implementation from 0.8 to reduce breaking changes when we port 0.8 to future versions, we will also use that for the me bridge in the near future
Some functions implemented for testing

SirEndii committed:
Merge branch 'dev/1.21.1' into dev/1.21.1-rs2

# Conflicts:
#	src/generated/resources/.cache/2f719b8031aed5e605ed8d6694b19245ca45c6ca
#	src/generated/resources/.cache/b8526e444ae7356037f3a813274f6835d1f3dd16
#	src/main/java/de/srendi/advancedperipherals/common/data/BlockStatesAndModelsProvider.java

SirEndii committed:
get and list items functions

SirEndii committed:
Initialize the network node container

SirEndii committed:
Merge branch 'dev/1.21.1' into dev/1.21.1-rs2

SirEndii committed:
Merge branch 'refs/heads/dev/1.21.1' into dev/1.21.1-rs2

SirEndii committed:
listFluids and getFluid
Bump RS

SirEndii committed:
Implement listCraftables functions

SirEndii committed:
Add patterns and filtered patterns support

SirEndii committed:
Implement `listDrives` and `listCells`

SirEndii committed:
Implement `getStoredEnergy` and `getEnergyCapacity`. Remove `getFilteredPatterns` and add the filter option to `getPatterns`

SirEndii committed:
Implement used, total and available storage functions for internal and external storages with chemical support

SirEndii committed:
implement import and export functions

SirEndii committed:
Added peripheral block support and make some inventory and fluid util methods nullable to support both at once

SirEndii committed:
Somewhat implemented the craftItems function, but it's not done yet. The items can be requested, but we are currently unable to give the players more information - already in discussion with the rs2 maintainer

SirEndii committed:
Change order of IStorageSystemPeripheral

SirEndii committed:
Basic chemical support

SirEndii committed:
Implemented all RS functions - the crafting job needs some love but I fixed that it would not have a crafting task to get information. Now needs major testing and refinement

SirEndii committed:
more chemical functions

SirEndii committed:
Chemical support for listing functions

SirEndii committed:
Full chemical support

SirEndii committed:
More cursed RS2 implementations but the bridge should be completed by now. We may change some api design before we release the update

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java:
```diff
@@ -110,6 +127,15 @@ public boolean cancel() {
+        if (craftingTask != null) {
+            for (TaskStatus status : autocraftingComponent.getStatuses()) {
+                if (status.info().id() == craftingTask.info().id()) {
+                    this.craftingTask = status;
```
```suggestion
                    this.craftingTask = status;
                    break;
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java:
```diff
@@ -134,11 +160,9 @@ public void tick() {
                 this.craftingTask = status;
-
                 // And only now we set that the crafting is started.
                 setStartedCrafting();
```
```suggestion
                setStartedCrafting();
                break;
```

SirEndii committed:
Update src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
Update src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

zyxkad commented on src/generated/resources/assets/advancedperipherals/lang/en_us.json:
```diff
@@ -59,6 +60,7 @@
   "item.advancedperipherals.tooltip.peripheral_casing": "&7An empty hull without the love it deserves. Used as a crafting ingredient",
   "item.advancedperipherals.tooltip.player_detector": "&7This peripheral can be used to interact with players, but don't be a stalker.",
   "item.advancedperipherals.tooltip.redstone_integrator": "&7This block is able to interact with redstone. Works exactly like the redstone api of an computer.",
+  "item.advancedperipherals.tooltip.rs_bridge": "&7The RS Bridge interacts with Refined Storage 2 to manage your items.",
```
idk if we need specifically mention `Refined Storage *2*` here, since below the id is `refinedstorage` not `refinedstorage2`

SirEndii replied:
Mhh yeah, you are prob right there

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,887 @@
+    }
+
+    @Override
+    @LuaFunction(mainThread = true)
```
does config reading have to run on main thread?
or I think `isEnabled` usually is not a Lua API here?

SirEndii replied:
Shouldn't be a lua function, not sure why I annoted it

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,887 @@
+    }
+
+    @Override
+    public MethodResult getChemical(IArguments arguments) throws LuaException {
```
missing `LuaFunction` I think? also I believe CC:T like you mark every `LuaFunction` with final keyword, since they should never be override.

SirEndii replied:
Good catch
There are also another functions with missing annotations and final keywords

SirEndii committed:
Merge branch 'refs/heads/dev/1.21.1' into dev/1.21.1-rs2

# Conflicts:
#	build.gradle
#	gradle.properties
#	src/main/java/de/srendi/advancedperipherals/common/addons/APAddons.java

SirEndii committed:
Merge remote-tracking branch 'origin/dev/1.21.1-rs2' into dev/1.21.1-rs2

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/StatusConstants.java:
```diff
@@ -0,0 +1,44 @@
+
+    private String info;
+
+    public StatusConstants withInfo(String info) {
```
not sure what are you trying to do here, are you sure it will be thread safe?

SirEndii replied:
Stolen from 0.8
An attempt on unified constants for return messages when something went wrong which we can document better
But tbf, I didn't implemented for everything yet on that branch

The withInfo method is meant to be used where some constants may need more information. For example "NOT_VALID_ITEM" where I would add the resource location of the item. 

For the thread safe part, I am not 100% sure. But to be honest, I am not an expert regarding thread safety. I don't see a reason why it should not be thread safe

SirEndii replied:
I will mark this as resolved and test at bit on the dev branch and implement the constants for more functions (I need the changes from this branch for the ME Bridge)

zyxkad replied:
withInfo sets the internal field of a constant instance, this may break if you do something like
```java
StatusConstants status_a = STATUS.withInfo("a");
StatusConstants status_b = STATUS.withInfo("b");
```
where later `status_a` and `status_b` will both returns `"b"`

zyxkad replied:
I think we should remove this method for clarity if its not used in 0.7, and I will check 0.8's code later

SirEndii replied:
aight

SirEndii replied:
Not sure why I completely overlooked that
What do you think about
```java
    public String withInfo(String extraInfo) {
        return this + "_" + extraInfo;
    }
``` 

instead

zyxkad replied:
yes this is perfect

SirEndii committed:
Add missing final keywords and lua function annotations

SirEndii committed:
Refined Storage 2 -> Refined Storage for the RS Bridge's description

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/IStorageSystemPeripheral.java:
```diff
@@ -0,0 +1,125 @@
+
+    MethodResult getEnergyUsage();
+
+    MethodResult getAvgPowerInjection();
```
I think `getAveragePowerInput` is a better name here, or `getAverageEnergyInput`.

SirEndii replied:
Yeah indeed
We copied that function name from ae2 directly, but I guess we should rename it

SirEndii committed:
getAvgPowerInjection -> getAverageEnergyInput

SirEndii committed:
Also rename the method that uses it

SirEndii committed:
Make StatusConstants.withInfo return a String directly

The `withInfo` method now directly returns the concatenation of the enum value and the provided info string, instead of attempting to store it internally.

Implementing this for anything is out of scope for this PR for now

SirEndii commented:
Everything should be right now 🙌🏻

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,893 @@
+        if (targetInventory == null)
+            return MethodResult.of(0, "Target Inventory does not exist");
+
+        return MethodResult.of(InventoryUtil.moveItem(targetInventory, itemHandler, filter.getLeft()), null);
```
do we need the second arguments as `null` here? imo it's redundancy.

same for other importToRSs

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,893 @@
+        if (!isAvailable())
+            return notConnected();
+
+        Pair<ItemFilter, String> filter = ItemFilter.parse(arguments.optTable(0, Map.of()));
```
will [`Collections.emptyMap()`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collections.html#emptyMap()) be better at here? so it doesn't create another map instance. Although its not a major impact.

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,893 @@
+            inventory = InventoryUtil.getHandlerFromName(computer, arguments.getString(1));
+
+        if (inventory == null)
+            return MethodResult.of(0, "The target inventory does not exist. Make sure the bridge is exposed in the computer network. Reach out to our discord or our documentation for help.");
```
I think the error message is too long here. maybe you want to put the hint sentence as the third return value, but still provide an easy identifiable error code as the second return value?

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,893 @@
+        if (!isAvailable())
+            return notConnected();
+
+        return MethodResult.of(0);
```
is this not available in RS api or is this a TODO thing

SirEndii replied:
There is a way in RS but it was not as straightforward so I skipped it. Gonna implement it later.
We need to loop through all nodes, and add the energy usage together

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,893 @@
+
+        GenericFilter<?> parsedFilter = filter.getLeft();
+
+        return MethodResult.of(RsApi.findPatternFromFilters(getNetwork(), null, parsedFilter).getLeft() != null);
```
I think `ifLeftPresent()` will be better than `getLeft() != null`

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,893 @@
+        if (hasInputFilter) {
+            Map<?, ?> inputFilterTable = TableHelper.getTableField(filterTable, "input");
+
+            inputFilter = GenericFilter.parseGeneric(inputFilterTable).getLeft();
```
did you check if the filter is valid or is it being ignored on purpose

SirEndii replied:
I didn't really think about it. It would return an empty filter by default so the function would not break
But I guess returning the debug message would be good

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java:
```diff
@@ -0,0 +1,251 @@
+
+    @Override
+    protected boolean isJobDone() {
+        return craftingTask != null && craftingTask.percentageCompleted() == 100;
```
idk if `percentageCompleted` returns an int or a double, but if it returns a double I'd recommend compare with `>= 100` and `< 100`, since float point numbers sometimes can have weird precise problems.

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java:
```diff
@@ -0,0 +1,251 @@
+            optionalPreview = futureCalculationResult.get();
+        } catch (ExecutionException | InterruptedException ex) {
+            AdvancedPeripherals.debug("Tried to get preview, but preview calculation is not done. Should be done.", org.apache.logging.log4j.Level.ERROR);
+            ex.printStackTrace();
```
I believe log4j's logger can directly log throwable, `printStackTrace` will never saves to log tho.

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsApi.java:
```diff
@@ -0,0 +1,679 @@
+     * @param filter  The filter here is optional, if an empty filter is provided, the method will return every resource
+     * @return a set of fluid stacks
+     */
+    public static Set<Map<String, Object>> getParsedCnemicals(Network network, ChemicalFilter filter) {
```
"Cnemicals"

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsApi.java:
```diff
@@ -0,0 +1,679 @@
+        return used;
+    }
+
+    public static long getTotalExternStorage(Network network, StorageType type) {
```
Extern -> External

same for belows

SirEndii replied:
do you think that should also be for lua functions?

zyxkad replied:
yeah sure

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsApi.java:
```diff
@@ -0,0 +1,679 @@
+        properties.put("disks", disks);
+        properties.put("mode", storageConfiguration.getFilterMode().toString());
+        //TODO Redstone mode
+        properties.put("redstone_mode", false);
```
If its a todo thing then I recommend we don't add this field right now? so people don't get confused when they saw a constant `false`

SirEndii replied:
ahh yeah, it is also not possible via rs to get that sate

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsChemicalHandler.java:
```diff
@@ -0,0 +1,50 @@
+            return resource;
+
+        ChemicalStack chemical = resource.copy();
+        chemical.setAmount(resource.getAmount() - component.insert(ChemicalResource.ofChemicalStack(chemical), resource.getAmount(), action == mekanism.api.Action.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY));
```
I recommend extract the inserted component here for better readability. same for below

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsFluidHandler.java:
```diff
@@ -0,0 +1,48 @@
+    public int fill(FluidStack resource, @NotNull FluidAction action) {
+        if (resource.isEmpty())
+            return 0;
+        return (int) (resource.getAmount() - component.insert(VariantUtil.ofFluidStack(resource), resource.getAmount(), action == FluidAction.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY));
```
same for here and below, make insert an separate line of statement since it changed some status.

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsItemHandler.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    public ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate) {
+        ItemStack inserted = stack.copy();
+        inserted.setCount((int) (stack.getCount() - component.insert(ItemResource.ofItemStack(stack), stack.getCount(), simulate ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY)));
```
same here, extract insert

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/blocks/blockentities/RsBridgeEntity.java:
```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    public void taskStatusChanged(@NotNull TaskStatus taskStatus) {
+        //jobs.stream().filter(job -> job.isCraftingStarted() && job.getCraftingTask().info().id().equals(taskStatus.info().id())).forEach(BasicCraftJob::jobStateChanged);
```
idk if this isn't working or is it WIP?

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/BasicCraftJob.java:
```diff
@@ -0,0 +1,206 @@
+        return debugMessage;
+    }
+
+    @LuaFunction(value = "getRequestedItem")
```
I believe you can ignore the `value =` if the only value is presenting

SirEndii replied:
I actually didn't know that haha

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/BasicCraftJob.java:
```diff
@@ -0,0 +1,206 @@
+        return cancel();
+    }
+
+    public abstract Object getParsedRequestedItem();
```
I will use `protected abstract` since other mods aren't & shouldn't use these.

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/BasicCraftJob.java:
```diff
@@ -0,0 +1,206 @@
+        return getParsedRequestedItem();
+    }
+
+    @LuaFunction(value = "getElapsedTime")
```
another way is rename `getElapsedTime` to `getElapsedTimeImpl` so we can use `getElapsedTime` as function name here and ignore the value param (which depends on if you prefer that).

SirEndii replied:
Mhh yeah, that is a better way

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/ChemicalFilter.java:
```diff
@@ -0,0 +1,162 @@
+        ChemicalFilter chemicalFilter = empty();
+        // If the map is empty, return a filter without any filters
+        if (item.isEmpty())
+            return Pair.of(chemicalFilter, null);
```
as a little optimization we can create a global EMPTY instance for this?

SirEndii committed:
Remove redundant null arguments

SirEndii committed:
Don't create new Map instances when using the default parameter for tables

SirEndii committed:
use present functions of Pair's instead

SirEndii committed:
Validate filters for `getPatterns`

SirEndii committed:
Improve debug message for import and export functions when no inventory was found

SirEndii committed:
Improve readability of RsChemicalHandler

SirEndii committed:
Remove redstone mode since it is not possible to get that state via the api

SirEndii committed:
Extract method calls from storage handlers to improve readability

SirEndii committed:
Uncomment forgotten jobStateChanged from our TaskStatusListener the bridge to listen for task changes

SirEndii committed:
Refactor how we call our methods for implementation and lua functions in our craft job

SirEndii committed:
Use a global EMPTY instance for our filters and don't always create a new empty instance

SirEndii committed:
Fix typo

SirEndii committed:
Print the craft job exceptions using log4j

SirEndii committed:
Change how we compare percentage completed of the rs craft job to prevent issues with floating point number precision

SirEndii committed:
Rename extern to external for our lua functions and implementatioons

SirEndii committed:
implement getEnergyUsage for the RS Bridge

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java:
```diff
@@ -0,0 +1,250 @@
+        // It's cursed, but the only way currently
+        if (craftingTask != null) {
+            for (TaskStatus status : autocraftingComponent.getStatuses()) {
+                if (status.info().id() == craftingTask.info().id()) {
```
are you sure `==` always works here? or do you think `equals` can be safer?

SirEndii replied:
I am 99% sure that it always uses the same instance
But to be safe, I changed it to equals

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSCraftJob.java:
```diff
@@ -0,0 +1,250 @@
+        PreviewType previewType = preview.type();
+
+        if (previewType == PreviewType.MISSING_RESOURCES) {
+            fireEvent(true, StatusConstants.MISSING_ITEMS);
+            calculationNotSuccessful = true;
```
```suggestion
            calculationNotSuccessful = true;
            fireEvent(true, StatusConstants.MISSING_ITEMS);
```

the order of `fireEvent` and `calculationNotSuccessful` is not same at here and below. Although they does not really matter I'd recommend make the order same to improve readability

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsApi.java:
```diff
@@ -0,0 +1,692 @@
+
+        GraphNetworkComponent graphNetworkComponent = network.getComponent(GraphNetworkComponent.class);
+        for (InWorldNetworkNodeContainer nodeContainer : graphNetworkComponent.getContainers(InWorldNetworkNodeContainer.class)) {
+            if (nodeContainer.getNode() instanceof StorageNetworkNode storageNetworkNode) {
+                CompositeStorage storage = (CompositeStorage) storageNetworkNode.getStorage();
+                for (Storage disk : storage.getSources()) {
+                    if (!(disk instanceof StateTrackedStorage stateTrackedStorage))
+                        continue;
+
+                    disks.add(parseStorageDisk(stateTrackedStorage));
+                }
+            }
```
```suggestion
            if (!(nodeContainer.getNode() instanceof StorageNetworkNode storageNetworkNode)) {
                continue;
            }
            CompositeStorage storage = (CompositeStorage) storageNetworkNode.getStorage();
            for (Storage disk : storage.getSources()) {
                if (disk instanceof StateTrackedStorage stateTrackedStorage) {
                    disks.add(parseStorageDisk(stateTrackedStorage));
                }
            }
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsApi.java:
```diff
@@ -0,0 +1,692 @@
+                type = "item";
+            } else if (serializableStorage.getType() == StorageTypes.FLUID) {
+                type = "fluid";
+            } else {
```
are we missing chemical here?

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/blocks/blockentities/RsBridgeEntity.java:
```diff
@@ -0,0 +1,130 @@
+    @Override
+    public void taskRemoved(@NotNull TaskId taskId) {
+        jobs.stream().filter(job -> job.isCraftingStarted() && job.getCraftingTask().info().id().equals(taskId)).forEach(BasicCraftJob::jobStateChanged);
+
```
```suggestion
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/StatusConstants.java:
```diff
@@ -0,0 +1,42 @@
+    JOB_DONE,
+    NOT_CRAFTABLE,
+    MISSING_ITEMS,
+    CPU_DOES_NOT_EXIST,
```
```suggestion
    CPU_NOT_FOUND,
```

I feel "does" is too much here

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/StatusConstants.java:
```diff
@@ -0,0 +1,42 @@
+    JOB_CANCELED,
+    JOB_DONE,
+    NOT_CRAFTABLE,
+    MISSING_ITEMS,
```
Is there any difference between `MISSING_ITEMS` and `ITEM_NOT_FOUND`?

I feel you need explain their usage (if they have different meaning) by comments at here.

SirEndii replied:
Described in the new commit
I will completely switch to these new constant in the next PR when this one is merged

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/BasicCraftJob.java:
```diff
@@ -0,0 +1,206 @@
+
+public abstract class BasicCraftJob {
+
+    protected static final String EVENT = "_crafting";
```
```suggestion
    private static final String EVENT = "_crafting";
```

I don't think `protected static` will ever work.

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/FluidFilter.java:
```diff
@@ -19,7 +24,9 @@
 import java.util.Map;
 
 //TODO tag
```
so what's the TODO here 🤔

SirEndii replied:
from the initial 1.21 port where I didn't implemented tag filtering
Forgot to remove it haha

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/GenericFilter.java:
```diff
@@ -0,0 +1,87 @@
+     * @return A pair of the parsed filter and an error message, if there is one
+     */
+    public static Pair<? extends GenericFilter<?>, String> parseGeneric(Map<?, ?> rawFilter) {
+        // TODO: Add chemical filter support
```
I think this is done or not?

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/ItemFilter.java:
```diff
@@ -95,19 +102,35 @@ public static Pair<ItemFilter, String> parse(Map<?, ?> item) {
+    }
+
+    @Override
+    public boolean testAE(GenericStack genericStack) {
```
have you tested if these functions can be load without AE or RS installed?

SirEndii replied:
you are right
Had it in mind to test that later, but I forgot to

SirEndii committed:
CPU_DOES_NOT_EXIST -> CPU_NOT_FOUND

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,905 @@
+    @Override
+    @LuaFunction(mainThread = true)
+    public final MethodResult isConnected() {
+        return MethodResult.of(isAvailable());
```
if only one value is always returned, I think it is not necessary to return a MethodResult

SirEndii committed:
Remove old todos

SirEndii committed:
Add comments to status constants

SirEndii committed:
protected static -> private static event field

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
remove empty line

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
change logic of listCells helper method

SirEndii committed:
Change order of fireEvent to improve readability

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
Add chemical type to storage disk property

SirEndii committed:
Change == to equals for TaskId check

SirEndii committed:
use boolean instead of MethodResult for isConnected() lua function

SirEndii committed:
Prevent peripheral creation issues when third party addons are not loaded like mekanism or the rs mekanism integration

SirEndii committed:
Improve todo comment on getTotalExternalStorage

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/GenericFilter.java:
```diff
@@ -0,0 +1,86 @@
+    // AE2 stuff
+    public abstract boolean testAE(GenericStack genericStack);
+
+    // AE2 stuff
```
```suggestion
    // RS stuff
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/GenericFilter.java:
```diff
@@ -0,0 +1,86 @@
+
+    public abstract boolean test(T toTest);
+
+    public static GenericFilter<?> empty() {
```
use constant here?

SirEndii committed:
begone import

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/GenericFilter.java:
```diff
@@ -0,0 +1,86 @@
+                    case "chemical":
+                        return ChemicalFilter.parse(rawFilter);
+                }
+            }
```
need return here or the next line will throw NPE

SirEndii replied:
The case will do nothing then and the method tries to find a valid filter using the name
I also committed a small fix now

This is a test using the isCraftable function
![image](https://github.com/user-attachments/assets/c3a2cae8-0113-46a3-a8d1-86fe81b5b4c9)

SirEndii committed:
Changelog

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,922 @@
+
+        for (RSCraftJob job : bridge.getJobs()) {
+            if (job.getId() == id) {
+                foundJob = job;
```
can break here

actually what I will do is directly return at here, and then return a empty methodresult after the for

SirEndii replied:
ahh yeah
totally no clue why I did it like that

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/ChemicalUtil.java:
```diff
@@ -0,0 +1,132 @@
+            return StringUtil.toHexString(md.digest(bytesOfHash));
+        } catch (NoSuchAlgorithmException ex) {
+            AdvancedPeripherals.debug("Could not parse fingerprint.", org.apache.logging.log4j.Level.ERROR);
+            ex.printStackTrace();
```
note

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/ChemicalUtil.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    public static ResourceLocation getRegistryKey(ChemicalStack fluid) {
+        return MekanismAPI.CHEMICAL_REGISTRY.getKey(fluid.copy().getChemical());
```
do we need copy to get chemical here?

SirEndii replied:
No, I copied it from mekanism and I didn't noticed it

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/FluidUtil.java:
```diff
@@ -30,6 +30,41 @@ public class FluidUtil {
+
+            return transferableAmount;
+        }
+
```
what if both inventory is not `IStorageSystemFluidHandler`? Is this situation not considered?

SirEndii replied:
Nope, that would be for example a RS System to RS System transfer which is not supported/AP does not offer that in any way

SirEndii committed:
AE2 Stuff -> RS Stuff

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
Remove unnecessary copy

SirEndii committed:
remove print stack trace and use log4j

SirEndii committed:
Fixed that filters can't specify a type while having a name property

SirEndii committed:
Use constant instead of creating a new object everytime for an empty generic filter

SirEndii committed:
Uncomment methods from testing and improve `getCraftingTask`

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,920 @@
+            inventory = InventoryUtil.getHandlerFromName(computer, arguments.getString(1));
+
+        if (inventory == null)
+            return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
```
I'd prefer nested if in this case, same for below. It improves readability.

```suggestion
        if (inventory == null) {
            inventory = InventoryUtil.getHandlerFromName(computer, arguments.getString(1));
            if (inventory == null) {
                return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
            }
        }
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,920 @@
+        if (!isAvailable())
+            return notConnected();
+
+        if (!APAddons.mekanismLoaded || !APAddons.refinedStorageMekanismLoaded)
```
I believe we only need to check `refinedStorageMekanismLoaded`, since it is not possible that RSM is loaded without mekanism. Same for other places

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsChemicalHandler.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public ChemicalStack extractChemical(ChemicalFilter filter, long count, mekanism.api.Action simulate) {
+        AdvancedPeripherals.debug("Trying to extract chemical from filter: " + filter);
```
I wonder if there is something like `debugf`, because filter will convert to string regardless if debug is enabled or not, which may cause performance issue.

SirEndii replied:
Indeed, but I am going to tackle that in another PR

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RsStorageTypes.java:
```diff
@@ -0,0 +1,25 @@
+
+    ITEM(StorageTypes.ITEM),
+    FLUID(StorageTypes.FLUID),
+    CHEMICAL(ChemicalResourceType.STORAGE_TYPE);
```
if mekanism is not loaded, can you use `ChemicalResourceType`?  It will be strange if `RsStorageTypes` can be success initialized.

SirEndii replied:
yeah it does not, not sure why I didn't thought about that.
But I am at a point where I am not sure how to handle that properly without spaghetti

edit: wait nvm I am stupid

SirEndii replied:
![image](https://github.com/user-attachments/assets/c6c7b8a4-2580-4e09-871f-f0efbc5affad)

yeah fixed it
No exceptions, no crashes

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/blocks/blockentities/MeBridgeEntity.java:
```diff
@@ -13,7 +13,7 @@
 import de.srendi.advancedperipherals.common.addons.appliedenergistics.CraftJob;
 import de.srendi.advancedperipherals.common.addons.appliedenergistics.MeBridgeEntityListener;
-import de.srendi.advancedperipherals.common.addons.computercraft.peripheral.MeBridgePeripheral;
+import de.srendi.advancedperipherals.common.addons.computercraft.peripheral.MEBridgePeripheral;
```
sometimes you are calling it `ME`, other times you are changing it to `Me`, so which one do we actually want lol

zyxkad replied:
so what I really want to say is the names are inconsistence. not only for ME bridge peripherals, but also for MeFluidHandler, and RS related stuffs.

to clarify, we need either make them both use capitalize letters (`ME` and `RS`) or both use one capitalize as the first letter and the second use lowercase letter (`Me` and `Rs`).

SirEndii replied:
I thought I renamed them all to RS and ME
I will check that tomorrow again

zyxkad replied:
> I thought I renamed them all to RS and ME I will check that tomorrow again

I see, that's git's problem on renaming cases.
you need do `git mv path/to/FILE tmp` and then `git mv tmp path/to/file` in order to record the case changes

SirEndii replied:
That's weird
Alright

SirEndii replied:
So, should be done

SirEndii committed:
rename Me bridge to ME bridge

SirEndii committed:
Switch to a nested if statement in inventory logic for better readability

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
remove useless mekanism check

SirEndii committed:
Fixed crash with storage functions when the mekanism addon is not loaded
Bump RS
Return 0 instead of empty list when mekanism addon is not loaded

SirEndii commented:
So everything done
One thing is, the functions go get available external storage is still not really supported, and I can't support it for a while, so that will stay like it is. I am going to place a hint in the docs for these functions
I need to discuss a bit with the RS maintainer

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,921 @@
+
+        IItemHandler inventory = InventoryUtil.getHandlerFromDirection(arguments.getString(1), owner);
+
+        if (inventory == null)
+            inventory = InventoryUtil.getHandlerFromName(computer, arguments.getString(1));
+
+        if (inventory == null)
+            return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
```
```suggestion
        if (inventory == null) {
            inventory = InventoryUtil.getHandlerFromName(computer, arguments.getString(1));
            if (inventory == null) {
                return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
            }
        }
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,921 @@
+            return notConnected();
+
+        IFluidHandler handler = FluidUtil.getHandlerFromDirection(arguments.getString(1), owner);
+        if (handler == null)
+            handler = FluidUtil.getHandlerFromName(computer, arguments.getString(1));
+
+        if (handler == null)
+            return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
```
```suggestion
        if (handler == null) {
            handler = FluidUtil.getHandlerFromName(computer, arguments.getString(1));
            if (handler == null) {
                return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
            }
        }
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,921 @@
+            return notConnected();
+
+        IFluidHandler handler = FluidUtil.getHandlerFromDirection(arguments.getString(1), owner);
+        if (handler == null)
+            handler = FluidUtil.getHandlerFromName(computer, arguments.getString(1));
+
+        if (handler == null)
+            return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
```
```suggestion
        if (handler == null) {
            handler = FluidUtil.getHandlerFromName(computer, arguments.getString(1));
            if (handler == null) {
                return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
            }
        }
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,921 @@
+            return MethodResult.of(Collections.emptyList());
+
+        IChemicalHandler handler = ChemicalUtil.getHandlerFromDirection(arguments.getString(1), owner);
+        if (handler == null)
+            handler = ChemicalUtil.getHandlerFromName(computer, arguments.getString(1));
+
+        if (handler == null)
+            return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
```
```suggestion
        if (handler == null) {
            handler = ChemicalUtil.getHandlerFromName(computer, arguments.getString(1));
            if (handler == null) {
                return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
            }
        }
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,921 @@
+            return MethodResult.of(Collections.emptyList());
+
+        IChemicalHandler handler = ChemicalUtil.getHandlerFromDirection(arguments.getString(1), owner);
+        if (handler == null)
+            handler = ChemicalUtil.getHandlerFromName(computer, arguments.getString(1));
+
+        if (handler == null)
+            return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
```
```suggestion
        if (handler == null) {
            handler = ChemicalUtil.getHandlerFromName(computer, arguments.getString(1));
            if (handler == null) {
                return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
            }
        }
```

SirEndii committed:
More nested if statements

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
More nested if statements

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
More nested if statements

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
More nested if statements

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
More nested if statements

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

SirEndii committed:
Rename Me -> ME and Rs -> RS

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,926 @@
+    @Override
+    @LuaFunction(mainThread = true)
+    public final MethodResult importChemical(IComputerAccess computer, IArguments arguments) throws LuaException {
+        if (!isAvailable())
+            return notConnected();
+
+        if (!APAddons.refinedStorageMekanismLoaded)
+            return MethodResult.of(Collections.emptyList());
+
+        IChemicalHandler handler = ChemicalUtil.getHandlerFromDirection(arguments.getString(1), owner);
+        if (handler == null) {
+            handler = ChemicalUtil.getHandlerFromName(computer, arguments.getString(1));
+            if (handler == null) {
+                return MethodResult.of(0, StatusConstants.INVENTORY_NOT_FOUND.name());
```
`notConnected()` is returning `nil, err`, `!refinedStorageMekanismLoaded` is returning an empty list, and `handler == null` is returning `0, err`.

I think we should make a standard that always return `nil, err` when an error happens, unless the action can be partially done.

This affects all import/export methods

SirEndii replied:
I agree
I changed it, so every method has a specific default value which will be documented 
The commit contains a description

SirEndii committed:
Bump CC to fix a runtime bug

SirEndii committed:
Add a custom default to the notConnected helper method.
And changed every "error" return value to a specific default value depending on the function:

Functions that return a number return 0.
Functions that return a collection return a final read only list.
Functions that return a boolean return false.
Functions that return an object return null.

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -286,11 +286,11 @@ public final MethodResult getChemical(IArguments arguments) throws LuaException
     public final MethodResult listItems(IArguments arguments) throws LuaException {
         if (!isAvailable())
-            return notConnected();
+            return notConnected(Collections.emptyList());
```
for lists, since it is not possible to return a half completed list, I'd recommend use `null` instead of emptyList. So user can check error easier, as well as its not creating extra lua tables

SirEndii replied:
I agree. But we should do that also for the other errors in these functions like these
https://github.com/IntelligenceModding/AdvancedPeripherals/blob/0d93afa17b1f82f2ee36f32bddc541ddf80b52c7/src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java#L293

zyxkad replied:
yes

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -788,7 +788,7 @@ public final MethodResult craftChemical(IComputerAccess computer, IArguments arg
     public final MethodResult getCraftingTasks() {
         if (!isAvailable())
-            return notConnected();
+            return notConnected(Collections.emptyList());
```
same for here

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -875,7 +874,7 @@ public final MethodResult isCrafting(IArguments arguments) throws LuaException {
     public final MethodResult getPatterns(IArguments arguments) throws LuaException {
         if (!isAvailable())
-            return notConnected();
+            return notConnected(Collections.emptyList());
```
and here as well as below

SirEndii committed:
Use null as a default return type for collections

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/blocks/blockentities/RsBridgeEntity.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class RsBridgeEntity extends PeripheralBlockEntity<RSBridgePeripheral> implements IPeripheralTileEntity, NetworkNodeContainerProvider, TaskStatusListener {
```
need rename this to `RSBridgeEntity`

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/blocks/blockentities/RsBridgeEntity.java:
```diff
@@ -0,0 +1,129 @@
+
+    @Override
+    public void taskStatusChanged(@NotNull TaskStatus taskStatus) {
+        jobs.stream().filter(job -> job.isCraftingStarted() && job.getCraftingTask().info().id().equals(taskStatus.info().id())).forEach(BasicCraftJob::jobStateChanged);
```
instead of use `&&`, we can chain up multiple filters, which may looks cleaner.

```suggestion
        jobs.stream().filter(BasicCraftJob::isCraftingStarted).filter(job -> job.getCraftingTask().info().id().equals(taskStatus.info().id())).forEach(BasicCraftJob::jobStateChanged);
```

same for `taskRemoved` and `taskAdded`

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/ChemicalUtil.java:
```diff
@@ -0,0 +1,131 @@
+
+        IChemicalHandler handler = extractHandler(location.getTarget(), null, null, null);
+        if (handler == null)
+            throw new LuaException("Target '" + name + "' is not a fluid handler");
```
```suggestion
            throw new LuaException("Target '" + name + "' is not a chemical handler");
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/util/inventory/ChemicalUtil.java:
```diff
@@ -0,0 +1,131 @@
+        return "";
+    }
+
+    public static ResourceLocation getRegistryKey(Chemical fluid) {
+        return MekanismAPI.CHEMICAL_REGISTRY.getKey(fluid);
+    }
+
+    public static ResourceLocation getRegistryKey(ChemicalStack fluid) {
+        return MekanismAPI.CHEMICAL_REGISTRY.getKey(fluid.getChemical());
```
```suggestion
    public static ResourceLocation getRegistryKey(Chemical chemical) {
        return MekanismAPI.CHEMICAL_REGISTRY.getKey(chemical);
    }

    public static ResourceLocation getRegistryKey(ChemicalStack chemical) {
        return MekanismAPI.CHEMICAL_REGISTRY.getKey(chemical.getChemical());
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSApi.java:
```diff
@@ -0,0 +1,694 @@
+        for (TrackedResourceAmount trackedResource : storage.getResources(Actor.EMPTY.getClass())) {
+            if (trackedResource.resourceAmount().resource() instanceof ItemResource itemResource && filter.test(itemResource.toItemStack())) {
+                items.add(getObjectFromItemResource(trackedResource.resourceAmount()));
+
```
random empty line catched XD
```suggestion
```

SirEndii replied:
F

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSApi.java:
```diff
@@ -0,0 +1,694 @@
+            boolean outputMatch = false;
+
+            if (inputFilter != null) {
+                outerLoop:
```
based on my knowledge on other languages labels usually has one less indent than the code its marking, but I'm not sure what is the common Java style. As long as you are happy with the looking I don't mind.

but besides, I suggest use full upper case for labels for easier identification.

SirEndii replied:
Mhh it seems that the style in modern java codebases is just like we have it currently but in caps

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSApi.java:
```diff
@@ -0,0 +1,694 @@
+                type = "item";
+            } else if (serializableStorage.getType() == StorageTypes.FLUID) {
+                type = "fluid";
+            } else if (serializableStorage.getType() == ChemicalResourceType.STORAGE_TYPE) {
```
again does this works without mekanism?

SirEndii replied:
It at least did in my testing
But I am gonna test this specific thing again

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSApi.java:
```diff
@@ -0,0 +1,694 @@
+            }
+            return getObjectFromFluidResource(new ResourceAmount(resource, count));
+        }
+        AdvancedPeripherals.debug("Could not create table from unknown resource " + resource.getClass() + " - Report this to the maintainer of ap", Level.WARN);
```
is there no `ChemicalResource` case? but it's used in `getCraftableChemicals`

SirEndii committed:
RsBridgeEntity -> RSBridgeEntity

SirEndii committed:
Chain stream filters in RSBridgeEntity for craft jobs

SirEndii committed:
fluid -> chemical in ChemicalUtil.java

SirEndii committed:
Remove empty line

SirEndii committed:
Fully upper case loop label

SirEndii committed:
Implement the actual support for chemical resources and fixed some crashes when the rs mekanism addon is not installed

SirEndii committed:
Fixed IllegalStateException when trying to get a holder from the registry, before we created a new holder instead of getting one from the registry

SirEndii committed:
Also return crafting tasks that weren't submitted by the Bridge
Add a crafted and resource attribute to the crafting task object

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/RSBridgePeripheral.java:
```diff
@@ -0,0 +1,925 @@
+            return notConnected(0);
+
+        if (!APAddons.refinedStorageMekanismLoaded)
+            return MethodResult.of(0);
```
there is no information about that why nothing is imported/exported

SirEndii replied:
a shit forgot about that

SirEndii committed:
add debug message to RS mekanism addon specific functions when the addon is not loaded

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSFluidHandler.java:
```diff
@@ -0,0 +1,50 @@
+    public RSFluidHandler(@NotNull Network network) {
+        this.network = network;
+        this.component = network.getComponent(StorageNetworkComponent.class);
+
```
```suggestion
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSApi.java:
```diff
@@ -0,0 +1,782 @@
+            if (nodeContainer.getNode() instanceof ExternalStorageNetworkNode storageNetworkNode) {
+                ExposedExternalStorage storage = (ExposedExternalStorage) storageNetworkNode.getStorage();
+
+                used += storage.getAll().stream().filter(amount -> type.getStorageType() != null && type.getStorageType().isAllowed(amount.resource())).mapToLong(ResourceAmount::amount).sum();
```
why don't we check `type.getStorageType() != null` in the beginning?

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSApi.java:
```diff
@@ -0,0 +1,782 @@
+        StorageConfiguration storageConfiguration = diskDrive.getStorageConfiguration();
+        CompositeStorage storage = (CompositeStorage) diskDrive.getStorage();
+
+        Set<Object> disks = new HashSet<>();
```
do we really need `Set` here instead of `List`?

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSChemicalHandler.java:
```diff
@@ -0,0 +1,53 @@
+        if (chemical == null)
+            return ChemicalStack.EMPTY;
+
+        ChemicalStack extracted = ChemicalUtil.toChemicalStack(chemical.chemical(), 1);
+        long amountExtracted = component.extract(chemical, filter.getCount(), simulate == mekanism.api.Action.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
+        extracted.setAmount(amountExtracted);
```
```suggestion
        long amountExtracted = component.extract(chemical, filter.getCount(), simulate == mekanism.api.Action.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
        ChemicalStack extracted = ChemicalUtil.toChemicalStack(chemical.chemical(), amountExtracted);
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSChemicalHandler.java:
```diff
@@ -0,0 +1,53 @@
+        if (resource.isEmpty())
+            return resource;
+
+        ChemicalStack inserted = resource.copy();
+        long amountInserted = component.insert(ChemicalResource.ofChemicalStack(inserted), resource.getAmount(), action == mekanism.api.Action.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
+        inserted.setAmount(resource.getAmount() - amountInserted);
+        return inserted;
```
```suggestion
        long amountInserted = component.insert(ChemicalResource.ofChemicalStack(inserted), resource.getAmount(), action == mekanism.api.Action.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
        ChemicalStack remain = resource.copyWithAmount(resource.getAmount() - amountInserted);
        return remain;
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSFluidHandler.java:
```diff
@@ -0,0 +1,50 @@
+        if (fluid == null)
+            return FluidStack.EMPTY;
+
+        FluidStack extracted = VariantUtil.toFluidStack(fluid, 1);
+        long amountExtracted = component.extract(fluid, filter.getCount(), simulate == FluidAction.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
+        extracted.setAmount((int) amountExtracted);
```
```suggestion
        long amountExtracted = component.extract(fluid, filter.getCount(), simulate == FluidAction.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
        FluidStack extracted = VariantUtil.toFluidStack(fluid, (int) amountExtracted);
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSItemHandler.java:
```diff
@@ -0,0 +1,54 @@
+    @NotNull
+    @Override
+    public ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate) {
+        ItemStack inserted = stack.copy();
+        long insertedAmount = component.insert(ItemResource.ofItemStack(stack), stack.getCount(), simulate ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
+        inserted.setCount((int) (stack.getCount() - insertedAmount));
+        return inserted;
```
```suggestion
        long insertedAmount = component.insert(ItemResource.ofItemStack(stack), stack.getCount(), simulate ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
        ItemStack remain = stack.copyWithCount((int) (stack.getCount() - insertedAmount));
        return remain;
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSItemHandler.java:
```diff
@@ -0,0 +1,54 @@
+        if (itemResource == null)
+            return ItemStack.EMPTY;
+
+        ItemStack extracted = itemResource.toItemStack();
+        long extractedAmount = component.extract(itemResource, count, simulate ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
+        extracted.setCount((int) extractedAmount);
```
```suggestion
        long extractedAmount = component.extract(itemResource, count, simulate ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
        ItemStack extracted = itemResource.toItemStack(extractedAmount);
```

zyxkad reviewed (changes requested):
some clean up

SirEndii committed:
Update src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSFluidHandler.java

Co-authored-by: Kevin Z <zyxkad@gmail.com>
Signed-off-by: Srendi <info@denizdemiroglu.de>

zyxkad committed:
clean up RS insert/extracts

Signed-off-by: Kevin Z <zyxkad@gmail.com>

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSChemicalHandler.java:
```diff
@@ -0,0 +1,51 @@
+        if (resource.isEmpty())
+            return resource;
+
+        long amountInserted = component.insert(ChemicalResource.ofChemicalStack(inserted), resource.getAmount(), action == mekanism.api.Action.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
```
```suggestion
        long amountInserted = component.insert(ChemicalResource.ofChemicalStack(resource), resource.getAmount(), action == mekanism.api.Action.SIMULATE ? Action.SIMULATE : Action.EXECUTE, Actor.EMPTY);
```

zyxkad committed:
fix insertChemical

Signed-off-by: Kevin Z <zyxkad@gmail.com>

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/refinedstorage/RSApi.java:
```diff
@@ -0,0 +1,782 @@
+     * @param filter  The filter here is optional, if an empty filter is provided, the method will return every resource
+     * @return a set of items
+     */
+    public static Set<Map<String, Object>> getParsedItems(Network network, ItemFilter filter) {
```
I really want a explanation about why are we using `Set` instead of `List` for these results.

Is it because storage may returns multiple same resources? or is it just a personal preference to use `Set` when possible.

SirEndii replied:
At the time where I started this, I somehow thought that Sets would be faster for just inserting and then iterating through it. But I was wrong, I researched that back then and Lists were actually faster which makes sense compared to a HashSet implementation. So I moved to Lists for the next functions but I never switched the functions which use Sets to a List - so I will do that next

zyxkad reviewed (approved):
Everything else looks good, at least for now :p

SirEndii commented:
Yeah I will address the two open reviews of you and then merge

SirEndii committed:
Switch to Lists instead of using HashSet implementations for the RS Bridge

SirEndii committed:
Move Storage Type check to the beginning of the function to save time
